### PR TITLE
Add basic support for loading pac file from file:// URI

### PR DESCRIPTION
**Pull request by k-f**
A quick attempt at file:// support per #12

### CODE REVIEW

**Comment by samuong** [comment]
Also, are you able to add a warning message (see issue #12 for context)? I'd suggest something like the following, but am open to word-smithing:

> Warning: The PAC URL is served over file://, which is supported by Alpaca, but not by Windows and macOS. Be careful if you configure your system settings to use the same URL.

**Comment by samuong** [comment]
Oh and finally, can you add a test for this? Just a function that creates a [temp file](https://godoc.org/io/ioutil#TempFile) with a simple FindProxyForURL pattern that just returns a fixed string like "PROXY proxy.example.com:80".. Then the test should check that `findProxyForRequest` returns a URL with the right host/port (i.e. proxy.exmaple.com:80).

**Comment by k-f** [comment]
@samuong thanks, changes done

File: proxyfinder.go [hunk 1]
```diff
@@ -88,7 +90,11 @@ func (pf *ProxyFinder) findProxyForRequest(req *http.Request) (*url.URL, error)
 }
 
 func (pf *ProxyFinder) downloadPACFile() {
-	resp, err := noProxyClient.Get(pf.pacURL)
+	if strings.HasPrefix(pf.pacURL, "file") {
```
Comment (samuong) [inline proxyfinder.go]: Change "file" to "file:" (with a colon)

File: proxyfinder_test.go [hunk 2]
```diff
@@ -72,3 +76,27 @@ func TestProxyWithoutPort(t *testing.T) {
 	pf := newProxyFinder(pacServer.URL, net.InterfaceAddrs)
 	checkProxyForURL(t, pf, "https://www.anz.com.au/", &url.URL{Host: "proxy.anz.com:80"})
 }
+
+func TestPacFromFilesystem(t *testing.T) {
+	//s := &proxyFinderTestServer{toAddrs("10.0.0.1"), "proxy.anz.com"}
```
Comment (samuong) [inline proxyfinder_test.go]: delete this comment

File: proxyfinder_test.go [hunk 3]
```diff
@@ -72,3 +76,27 @@ func TestProxyWithoutPort(t *testing.T) {
...
+
+	content := []byte("function FindProxyForURL(url, host) { return \"PROXY proxy.example.com:80\"}")
```
Comment (samuong) [inline proxyfinder_test.go]: use a ` for the outer "s, and an unescaped " for the inner ones

File: proxyfinder_test.go [hunk 4]
```diff
@@ -72,3 +76,27 @@ func TestProxyWithoutPort(t *testing.T) {
...
+	tmpfile, err := ioutil.TempFile("", "test.pac")
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (samuong) [inline proxyfinder_test.go]: you can use `require.Nil(t, err)` instead of the if statement

File: proxyfinder_test.go [hunk 5]
```diff
@@ -72,3 +76,27 @@ func TestProxyWithoutPort(t *testing.T) {
...
+
+	defer os.Remove(tmpfile.Name())
+
+	if _, err := tmpfile.Write(content); err != nil {
+		log.Fatal(err)
+	}
+	if err := tmpfile.Close(); err != nil {
+		log.Fatal(err)
+	}
```
Comment (samuong) [inline proxyfinder_test.go]: Again, you can use require.Nil here:

```
require.Nil(t, tmpfile.Write(content))
require.Nil(t, tmpfile.Close())
```

This should make the function concise enough that you can get rid of all the blank lines.

File: proxyfinder.go [hunk 6]
```diff
@@ -16,28 +16,30 @@ import (
...
 }
 
 func newProxyFinder(pacURL string, getAddrs addressProvider) *ProxyFinder {
-	pf := &ProxyFinder{pacURL: pacURL, netMonitor: NewNetMonitor(getAddrs)}
+	transport := &http.Transport{Proxy: nil}
+	transport.RegisterProtocol("file", http.NewFileTransport(http.Dir("/")))
+	noProxyClient := &http.Client{
+		Transport: transport,
+		Timeout:   30 * time.Second,
+	}
```
[the remark this answers is not in the document]
Comment (k-f) [reply proxyfinder.go]: @samuong  - I haven't done this as I have realised that using `file://` breaks the online / offline checks.  Instead I've added a section where if `file://` is used for the PAC file the proxy finder tries to directly access google.com, if the direct access works then it sets the proxy as offline.   To do this check a client that supports both file and direct http is needed.

[the remark this answers is not in the document]
Comment (samuong) [reply proxyfinder.go]: Sorry, I didn't quite understand how it breaks the online/offline checks. The check is fully within `ProxyFinder.downloadPACFile()`, and for file:// URLs it should basically always succeed, given that you wouldn't normally have network errors between Alpaca and the filesystem. Let me know if I'm not correct here?

I'm not sure if there's some confusion due to naming, and I have to admit I named this variable really badly when I wrote it. The word `online` actually means that Alpaca has connectivity _to the PAC server_. If you think about this from the perspective of someone on a corporate network, when they have access to the PAC server, they often _don't_ have access to google.com. I've got a fix for this that's part of a larger change I'm making, which I'll hopefully finish off soon.

[the remark this answers is not in the document]
Comment (samuong) [reply proxyfinder.go]: I just had a chance to take another look at this, and I think I understand what you're saying now. Basically you've got a PAC file which refers to proxies that are on an internal network, and the PAC download will always succeed, even though the proxies won't be resolvable when we're on a public Internet connection?

In that case what I think it'd be better to do a reverse lookup of 8.8.8.8 and 2001:4860:4860::8888, and if _either_ of them succeed, then assume we're on the public Internet, and we don't need to use the PAC file. There are a few benefits of this:

1. A (reverse) DNS lookup is going to be faster than a full HTTP request (which might get stuck on a slow network connection)
2. You don't need an HTTP Transport, so the file-only Transport will do

You could use `net.LookupAddr()` for this:

```
_, err1 := net.LookupAddr("8.8.8.8")
_, err2 := net.LookupAddr("2001:4860:4860::8888")
if err1 == nil || err2 == nil {
    pf.online = false
    return
}
```

Thoughts?

File: proxyfinder.go [hunk 7]
```diff
@@ -111,6 +117,23 @@ func (pf *ProxyFinder) downloadPACFile() {
...
 		return
 	}
+
+	if strings.HasPrefix(pf.pacURL, "file:") {
+		// When using a local PAC file the online/offline status can't be detirmined by the fact the PAC file is returned
+		// Instead try reverse DNS resolution of an internet address
+		const timeout = 2 * time.Second
+		ctx, cancel := context.WithTimeout(context.TODO(), timeout)
+		defer cancel() // important to avoid a resource leak
+		var r net.Resolver
```
[the remark this answers is not in the document]
Comment (samuong) [reply proxyfinder.go]: Actually I just realised that Go's DNS resolution function doesn't take a timeout, so if you're building your own context just for this, then please ignore my comment here.

[the remark this answers is not in the document]
Comment (k-f) [reply proxyfinder.go]: Yep that was the exact reason for this. Without the custom context in some circumstances it ended up causing things to hang for ages when switching network locations.

**Pull request merged**